framework/src/messaging/messaging_recv.c

```diff
@@ -0,0 +1,296 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+#include <tinyara/config.h>
+#include <debug.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <mqueue.h>
+#include <stdlib.h>
+#include <signal.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/types.h>
+#include <messaging/messaging.h>
+#include "messaging_internal.h"
+
+/****************************************************************************
+ * private functions
+ ****************************************************************************/
+/****************************************************************************
+ * Name : set_notify_signal
+ * 
+ * Description:
+ *  This function calls mq_notify to set message queue notification.
+ ****************************************************************************/
+static int set_notification(messaging_recv_info_t *data)
+{
+	int ret;
+	struct sigevent notification;
+
+	notification.sigev_notify = SIGEV_SIGNAL;
+	notification.sigev_signo = SIGMSG_UNICAST;
+	notification.sigev_value.sival_ptr = data;
+
+	ret = mq_notify(data->mqdes, &notification);
+	if (ret < 0) {
+		msgdbg("message recv fail : set notification fail. %d\n", errno);
+		MSG_FREE(data);
+		mq_close(data->mqdes);
+		return ERROR;
+	}
+	return OK;
+}
+
+/****************************************************************************
+ * Name : wrapper_callback
+ * 
+ * Description:
+ *  This function is wrapper callback function for waiting until receiving the message.
+ ****************************************************************************/
+static void wrapper_callback(int signo, siginfo_t *data)
+{
+	int ret;
+	messaging_recv_info_t *info;
+	char msg[CONFIG_MESSAGING_UNICAST_MSGSIZE];
+	struct mq_attr attr;
+
+	if (data == NULL) {
+		msgdbg("message recv fail : wrong param for wrapper callback(data).\n");
+		return;
+	}
+
+	info = (messaging_recv_info_t *)data->si_value.sival_ptr;
+	if (info == NULL) {
+		msgdbg("message recv fail : wrong param for wrapper callback(info).\n");
+		return;
+	}
+
+	while (1) {
+		ret = mq_receive(info->mqdes, (char *)&msg, sizeof(messaging_recv_info_t), 0);
+		if (ret < 0) {
+			msgdbg("message recv fail : mq_receive fail. %d\n", errno);
+			mq_close(info->mqdes);
+			MSG_FREE(data);
+			return;
+		}
+		/* Call user callback */
+		(*info->user_cb)(info->cb_data, &msg);
+
+		ret = mq_getattr(info->mqdes, &attr);
```

Comment:
of course it set attributes before send, but mq_receive or mq_timedreceive returns same as msgsize.
so you can check validation by check size of msg and ret, see below
```
ssize_t ret = mq_receive();
if (ret != sizeof(msg)) {
    do_errors;
}
```

Reply:
Done.